### ISSUE TITLE
How to setup SMTP if one knows the access token already?

### ISSUE THREAD

**Issue by simurgh9** [harishkrupo/oauth2ms#10]
I obtained the token using instructions given [here](https://github.com/UvA-FNWI/M365-IMAP). I am able to download email with it. I tried to adapt the code you give for SMTP to use that token. I came up with this,

```lisp
   (use-package smtpmail
      :ensure t
      :config
      (add-to-list 'smtpmail-auth-supported 'xoauth2)
      :init
      (cl-defmethod smtpmail-try-auth-method
        (process (_mech (eql xoauth2)) user password)
        (let* ((access-token "<token string here"))
          (smtpmail-command-or-throw
           process
           (concat "AUTH XOAUTH2 " access-token)
           235)))
      (setq message-send-mail-function 'smtpmail-send-it
            message-kill-buffer-on-exit t
            starttls-use-gnutls t
            smtpmail-default-smtp-server "smtp.example.com"
            smtpmail-smtp-server "smtp.example.com"
            smtpmail-stream-type  'starttls
            smtpmail-starttls-credentials
            '(("smtp.example.com" 587 nil nil))
            smtpmail-auth-credentials
            '(("smtp.example.com" 587 "emailaddress@example.com" nil))
            smtpmail-smtp-service 587))
```

When sending the email, the authentication still fails. Any ideas?

**Comment by harishkrupo** [comment]
From a quick glance, it looks like the tool returns raw tokens. However you need to use xoauth encoded tokens for SMTP. The `--encode-xoauth2` option in the `oauth2ms` tool does exactly that. 
You can either adapt that code in elisp or simply use the `oauth2ms` tool :)

**Comment by simurgh9** [comment]
I think the issue before that is that I was confusing refresh token with access token. If my understanding is right, refresh token is used to fetch the access token right?

**Comment by harishkrupo** [comment]
Yes, that is correct. https://auth0.com/learn/refresh-tokens/
That said, I am not sure how you are receiving refresh tokens. AFAIK the msal library internally handles refresh tokens and doesn't provide them to the user.
Could you please explain a bit more about what you are trying? I might be able to help if I know more :)

**Comment by simurgh9** [comment]
Alright, I figured it out. My refresh token was not asked for with scope `https://outlook.office.com/SMTP.Send`. I also was not encoding per [this](https://docs.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#sasl-xoauth2). I fixed those two and now I am able to send mail. If anyone wants the Elisp I wrote, reopen this issue and I'll put it here. It is just a couple of functions.
